verify.js

```diff
@@ -85,97 +62,135 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
-    }
-    if (payload.nbf > clockTimestamp + (options.clockTolerance || 0)) {
-      return done(new NotBeforeError('jwt not active', new Date(payload.nbf * 1000)));
-    }
-  }
-
-  if (typeof payload.exp !== 'undefined' && !options.ignoreExpiration) {
-    if (typeof payload.exp !== 'number') {
-      return done(new JsonWebTokenError('invalid exp value'));
-    }
-    if (clockTimestamp >= payload.exp + (options.clockTolerance || 0)) {
-      return done(new TokenExpiredError('jwt expired', new Date(payload.exp * 1000)));
-    }
-  }
-
-  if (options.audience) {
-    var audiences = Array.isArray(options.audience)? options.audience : [options.audience];
-    var target = Array.isArray(payload.aud) ? payload.aud : [payload.aud];
-
-    var match = target.some(function(targetAudience) {
-      return audiences.some(function(audience) {
-        return audience instanceof RegExp ? audience.test(targetAudience) : audience === targetAudience;
-      });
-    });
-
-    if (!match)
-      return done(new JsonWebTokenError('jwt audience invalid. expected: ' + audiences.join(' or ')));
-  }
-
-  if (options.issuer) {
-    var invalid_issuer =
-        (typeof options.issuer === 'string' && payload.iss !== options.issuer) ||
-        (Array.isArray(options.issuer) && options.issuer.indexOf(payload.iss) === -1);
-
-    if (invalid_issuer) {
-      return done(new JsonWebTokenError('jwt issuer invalid. expected: ' + options.issuer));
-    }
-  }
-
-  if (options.subject) {
-    if (payload.sub !== options.subject) {
-      return done(new JsonWebTokenError('jwt subject invalid. expected: ' + options.subject));
-    }
-  }
-
-  if (options.jwtid) {
-    if (payload.jti !== options.jwtid) {
-      return done(new JsonWebTokenError('jwt jwtid invalid. expected: ' + options.jwtid));
-    }
-  }
-
-  if (options.maxAge) {
-    if (typeof payload.iat !== 'number') {
-      return done(new JsonWebTokenError('iat required when maxAge is specified'));
-    }
-
-    var maxAgeTimestamp = timespan(options.maxAge, payload.iat);
-    if (typeof maxAgeTimestamp === 'undefined') {
-      return done(new JsonWebTokenError('"maxAge" should be a number of seconds or string representing a timespan eg: "1d", "20h", 60'));
-    }
-    if (clockTimestamp >= maxAgeTimestamp + (options.clockTolerance || 0)) {
-      return done(new TokenExpiredError('maxAge exceeded', new Date(maxAgeTimestamp * 1000)));
-    }
-  }
-
-  return done(null, payload);
+  var getSecret;
+
+  if(typeof secretOrPublicKey === 'function') {
+      if(!callback) {
+          return done(new JsonWebTokenError('verify must be called asynchronous if secret or public key is provided as a callback'));
+      }
+
+      getSecret = secretOrPublicKey;
+  }
+  else {
+      getSecret = function(header, secretCallback) {
+          return secretCallback(null, secretOrPublicKey);
+      };
+  }
+
+  return getSecret(header, function(err, secretOrPublicKey) {
```

Comment:
isn't enough with the `kid`? What's on the `header` that the user might want to use for key retrieval?

Reply:
You may want to use other headers to find the key: https://tools.ietf.org/html/rfc7515#page-10